### DISCUSSION TITLE
Rolling back a committed transaction

### DISCUSSION THREAD

**Discussion by pathange-s** [go-gorm/gorm#7265] [Q&A]
How can I rollback a committed transaction?

I am working on a simple two phase commit:
1. Write on api db with apiTx
2. Write on cdpc db with cdpcTx
```
	// Commit Phase - try committing both transactions
	if err := apiTx.Commit().Error; err != nil {
		// If committing apiTx fails, rollback both transactions
		rollback()
		return errors.InternalServerError(errors.ErrInternalServer, "error committing api transaction: "+err.Error())
	}

	if err := cdpcTx.Commit().Error; err != nil {
		// If cdpcTx commit fails after apiTx has already committed, attempt to "compensate" by reversing apiTx
		// Need to check if we can rollback a committed transaction
		apiRollbackErr := apiTx.Rollback().Error
		if apiRollbackErr != nil {
			return errors.InternalServerError(errors.ErrInternalServer, "error committing cdpc transaction and failed to rollback api transaction: "+err.Error()+"; "+apiRollbackErr.Error())
		}
		return errors.InternalServerError(errors.ErrInternalServer, "error committing cdpc transaction after api commit: "+err.Error())
	}
```

**Comment by hexvalid** [comment]
You cannot roll back `apiTx` after its `Commit()` has succeeded. GORM's SQL transaction is backed by `database/sql.Tx`, which is finished after commit or rollback; further transaction operations return `sql.ErrTxDone`. See [sql.Tx](https://pkg.go.dev/database/sql#Tx).

I checked this with GORM v1.31.2 and SQLite: insert a row inside a transaction, commit, then call `Rollback()`. Rollback returned `sql.ErrTxDone`, and the committed row remained.

The two sequential `Commit()` calls in the example do not implement two-phase commit. That protocol requires participants to prepare before the coordinator decides the final outcome. For example, PostgreSQL exposes [PREPARE TRANSACTION](https://www.postgresql.org/docs/current/sql-prepare-transaction.html), but using it requires appropriate database support and a coordinator that can recover unfinished transactions.

For your failure path, reversing an already committed change would need a **new compensating transaction** implementing the business-specific inverse operation. That operation also needs durable retry/idempotency handling, and it cannot guarantee that nobody observed the intermediate state.

If both writes can be performed through one database transaction, use one GORM transaction for both. If they must stay in independent databases, choose a supported distributed-transaction mechanism or an explicit compensation/outbox workflow according to the consistency you need. Calling `apiTx.Rollback()` after the successful commit cannot restore atomicity.
